packages/react-reconciler/src/ReactFiberHooks.js

```diff
@@ -837,6 +853,282 @@ function rerenderReducer<S, I, A>(
+  } else {
+    invariant(
+      false,
+      'Cannot read from mutable source during the current render without tearing. This is a bug in React. Please file an issue.',
+    );
+  }
+}
+
+type MutableSourceRef<Source, Snapshot> = {|
+  getSnapshot: null | MutableSourceGetSnapshotFn<Source, Snapshot>,
+  source: null | MutableSource<Source>,
+  subscribe: null | MutableSourceSubscribeFn<Source, Snapshot>,
+|};
+
+function useMutableSourceImpl<Source, Snapshot>(
+  useEffectImpl: (
+    create: () => (() => void) | void,
+    deps: Array<mixed> | void | null,
+  ) => void,
+  useRefImpl: (
+    initialValue: MutableSourceRef<Source, Snapshot>,
+  ) => {|current: MutableSourceRef<Source, Snapshot>|},
+  useStateImpl: (
+    initialState: () => MutableSourceState<Source, Snapshot>,
+  ) => [MutableSourceState<Source, Snapshot>, Dispatch<any>],
+  source: MutableSource<Source>,
+  getSnapshot: MutableSourceGetSnapshotFn<Source, Snapshot>,
+  subscribe: MutableSourceSubscribeFn<Source, Snapshot>,
+): Snapshot {
+  const getVersion = source._getVersion;
+  const version = getVersion();
+
+  const [state, setState] = useStateImpl(
+    () =>
+      ({
+        getSnapshot,
+        snapshot: readFromUnsubcribedMutableSource(source, getSnapshot),
+        source,
+        subscribe,
+      }: MutableSourceState<Source, Snapshot>),
+  );
+
+  const ref = useRefImpl({
+    getSnapshot: null,
+    source: null,
+    subscribe: null,
+  });
+
+  let didUnsubscribe = false;
+
+  // If we got a new source or subscribe function,
+  // we'll need to subscribe in a passive effect,
+  // and also check for any changes that fire between render and subscribe.
+  useEffectImpl(() => {
+    const shouldResubscribe =
+      subscribe !== ref.current.subscribe || source !== ref.current.source;
+
+    // We need to re-subscribe any time either the source or the subscribe function changes.
+    // We dont' need to re-subscribe when getSnapshot changes,
+    // but we will need the latest getSnapshot inside of our change handler-
+    // so we store a reference to it the ref as well.
+    ref.current = {
+      getSnapshot,
+      source,
+      subscribe,
+    };
+
+    if (shouldResubscribe) {
+      const handleChange = () => {
+        // It's possible that this callback will be invoked even after being unsubscribed,
+        // if it's removed as a result of a subscription event/update.
+        // In this case, React might log a DEV warning about an update from an unmounted component.
+        // We can avoid triggering that warning with this check.
+        if (didUnsubscribe) {
+          return;
+        }
+
+        const latestGetSnapshot = ((ref.current
+          .getSnapshot: any): MutableSourceGetSnapshotFn<Source, Snapshot>);
+        let newSnapshot = null;
+        let snapshotDidThrow = false;
+        try {
+          newSnapshot = latestGetSnapshot(source._source);
+        } catch (error) {
+          snapshotDidThrow = true;
```

Comment:
What's the benefit of this change?

Happy to make the change. Just curious.

Reply:
Main benefit is not passing a closure in the case where there's no error. If you needed to pass a closure anyway (like to check for stale sources, which I don't think you do for the reasons described in my earlier comment), then this would still have the benefit of saving one extra check from the common path.